### DISCUSSION TITLE
Gateway sends join-acept but device activateOTAA() returns -1116 anyway

### DISCUSSION THREAD

**Discussion by quintesse** [jgromes/RadioLib#1341] [Q&A]
Given the title what could I be doing wrong?

It seems to me that the RadioLib configuration should be at least correct enough to send the join request and get it accepted. I also haven't noticed any packet loss (device and gateway are in the same room) so I'm assuming it's not that the device isn't receiving the reply.

The code I'm using is functionally the same as LoRaWAN_Starter.

The Live Data in the console shows "js.join.accept", followed by "ns.up.join.process" and finally a "as.up.join.forward".

I haven't added any logs but can provide them.

**Comment by StevenCellist** [comment]
Please provide your hardware setup, software setup and the (debug protocol) log of a full join cycle, with millisecond timestamps enabled in your terminal.

**Reply by quintesse** [reply]
The hardware is a XIAO ESP32S3 with a Wio-SX1262 LoRa module.
Not sure what you want to know about the software setup, a link to a code repo?

And I have some logs but they are without timestamps, sorry, I'd have to figure out how to configure that 😅 

```
12:40:04.738 > INF: Initialise the LoRaWan radio...
12:40:04.788 > INF: Joining the LoRaWAN Network...
12:40:04.788 > RLB_PRO: Setting up dynamic channels
12:40:04.788 > RLB_PRO: UL:   0 1 868.100 (0 - 5) | DL:   0 1 868.100 (0 - 5)
12:40:04.788 > RLB_PRO: UL:   1 1 868.300 (0 - 5) | DL:   1 1 868.300 (0 - 5)
12:40:04.789 > RLB_PRO: UL:   2 1 868.500 (0 - 5) | DL:   2 1 868.500 (0 - 5)
12:40:04.789 > RLB_PRO: [MAC] 0x03
12:40:04.789 > RLB_PRO: 00000000: 20
12:40:04.789 > RLB_PRO: LinkAdrReq: dataRate = 2, txSteps = 0, nbTrans = 0
12:40:04.790 > RLB_PRO: LinkAdrAns: 07
12:40:04.790 > RLB_PRO: [MAC] 0x04
12:40:04.790 > RLB_PRO: 00000000: 07                                               .
12:40:04.790 > RLB_PRO: DutyCycleReq: max duty cycle = 1/2^7
12:40:04.790 > RLB_PRO: [MAC] 0x05
12:40:04.790 > RLB_PRO: 00000000: 00 d2 ad 84                                      ....
12:40:04.790 > RLB_PRO: RXParamSetupReq: Rx1DrOffset = 0, rx2DataRate = 0, freq = 869.525
12:40:04.797 > RLB_PRO: [MAC] 0x08
12:40:04.797 > RLB_PRO: 00000000: 01                                               .
12:40:04.797 > RLB_PRO: RXTimingSetupReq: delay = 1 sec
12:40:04.797 > RLB_PRO: [MAC] 0x09
12:40:04.797 > RLB_PRO: 00000000: 05                                               .
12:40:04.797 > RLB_PRO: [MAC] 0x0c
12:40:04.797 > RLB_PRO: 00000000: 65                                               e
12:40:04.797 > RLB_PRO: ADRParamSetupReq: limitExp = 6, delayExp = 5
12:40:04.797 > RLB_PRO: [MAC] 0x0f
12:40:04.797 > RLB_PRO: 00000000: fa                                               .
12:40:04.797 > RLB_PRO: RejoinParamSetupReq: maxTime = 15, maxCount = 10
12:40:04.867 > RLB_PRO: 
12:40:04.867 > RLB_PRO: PHY:  Frequency = 868.300 MHz, TX = 16 dBm
12:40:04.868 > RLB_PRO: LoRa: SF = 10, BW = 125.0 kHz, CR = 4/5, IQ: U
12:40:05.246 > RLB_PRO: JoinRequest sent (DevNonce = 0) <-- Rx Delay start
12:40:05.247 > RLB_PRO: 00000000: 00 00 00 00 00 00 00 00 00 06 d0 93 c0 7e d5 b3  .............~..
12:40:05.247 > RLB_PRO: 00000010: 70 00 00 87 ca 84 cf                             p......
12:40:05.248 > RLB_PRO:
12:40:05.248 > RLB_PRO: PHY:  Frequency = 868.300 MHz, TX = 16 dBm
12:40:05.249 > RLB_PRO: LoRa: SF = 10, BW = 125.0 kHz, CR = 4/5, IQ: D
12:40:10.243 > RLB_PRO: Opening Rx1 window (231 ms timeout)... <-- Rx Delay end 
12:40:10.474 > RLB_PRO: Closing Rx1 window
12:40:10.475 > RLB_PRO:
12:40:10.475 > RLB_PRO: PHY:  Frequency = 869.525 MHz, TX = 16 dBm
12:40:10.476 > RLB_PRO: LoRa: SF = 12, BW = 125.0 kHz, CR = 4/5, IQ: D
12:40:11.243 > RLB_PRO: Opening Rx2 window (688 ms timeout)... <-- Rx Delay end 
12:40:11.931 > RLB_PRO: Closing Rx2 window
12:40:11.931 > ERR: Failed to join LoRaWan network: #-1116
```

**Reply by StevenCellist** [reply]
> Not sure what you want to know about the software setup, a link to a code repo?

Preferably, a dump here of the first part of `config.h` (redact AppKey/NwkKey, the EUIs can be intercepted anyway) up until `LoRaWANNode node...` as well as your main file up until `node.activateOTAA()`. If you use ArduinoIDE, it comes with a builtin button for timestamps; if you use PlatformIO, you can add timestamps using `monitor_filters = time` in your .ini. Otherwise, check documentation of your terminal program.

**Reply by quintesse** [reply]
I've updated the log with a new one with timings and the code can be found here: https://github.com/quintesse/arduino/blob/main/ESP32S3/tof_oled_lorawan/src/main.cpp

**Comment by HeadBoffin** [comment]
> The code I'm using is functionally the same as LoRaWAN_Starter.

The devil being in the details, perhaps share that or use LW Starter.


> device and gateway are in the same room

Tiny device says "Knock knock", giant responds in booming voice "You're in!", device totally overwhelmed with the volume & doesn't understand anything. LoRa = Long Range - 5m & a brick wall separation is a typical minimum - if your JR RSSI is -50 or above it will need moving.

TL;DR: You've got known good hardware, the MCU & radio are talking, so this looks likes the classic "too close" problem.

**Reply by quintesse** [reply]
Ok, I honestly didn't know a signal could be too strong. Is there a way for me to see this JR RSSI? (And I'm guessing that "above" here means "> -50"?)

**Reply by StevenCellist** [reply]
Assuming you use TTN, when you click on one of the JoinRequest entries (the middle one of the three), you will see the RSSI in the `rx_metadata` field. And indeed, values >-50 are usually problematic.

**Reply by HeadBoffin** [reply]
Technically it's < -50 that's a problem - but just ignore the - sign and think bigger than ...

Or is it the other way round. Who decided it should be a minus number. Meh. All I knows is -70 is good and -120 is getting a bit weak.

**Reply by quintesse** [reply]
Aaah ok! That might actually be the issue then. I think I'll have to check the device antenna connection then, because looking at the TTN logs I can see -130, so that's probably on the edge of being too low. Is there a way to see the RSSI of signals received by the device? (In the debug protocol log for example)

**Reply by StevenCellist** [reply]
Well.. it's hard to print out the strength of a signal that it doesn't receive, huh? ;)
But -130 looks like at least one of not both ends of the link have trouble with their antenna.

**Reply by quintesse** [reply]
Hehe true, but I was hoping that if _something_ came in there would be at least be a way to see if the signal strength was okay or not :-)

**Reply by StevenCellist** [reply]
Check the Reference example: it shows how to print the RSSI and SNR of a received packet.

**Reply by quintesse** [reply]
Ok, given the fact that the `getRSSI()` (and `getSNR()`) is on `radio` can I assume that it would return a value for any packets that are received that are not passed on to user code (like I assume would happen for "system" packets like accepting a join request)? Or am I being too optimistic? :-)

**Reply by StevenCellist** [reply]
That is too optimistic:  there is no such thing as 'system packets' that you wouldn't see. Any and all LoRaWAN communication shows up in the debug protocol logs. Really, just triple check your antenna on device and gateway and make sure there's 10 meter and a brick wall in between both of them. That is very very likely to resolve this problem.

**Reply by quintesse** [reply]
Yeah sorry, my question was just out of curiosity :-) That's why I was thinking there must be some way the system knows that an `activateOTAA()` failed, which would be because the gateway sent an error packet.

**Reply by StevenCellist** [reply]
The reason why a system (= the device) knows it failed, is because it didn't receive the JoinAccept, which is a necessity. LoRaWAN doesn't do such thing as an 'error packet'.

**Reply by quintesse** [reply]
Oh? So how are errors like "DevNonce is too small" handled? Wouldn't that require the gateway to send some kind of failure? (I understand perfectly if you simply say: RTFM 😉 )

**Reply by HeadBoffin** [reply]
RTFM

**Reply by HeadBoffin** [reply]
But because this is adding to my detailed files for training notes, I'll be a little kinder. But only a little.

Once upon a time an old fa&t wrote the [notes.md](https://github.com/jgromes/RadioLib/blob/master/examples/LoRaWAN/LoRaWAN_Starter/notes.md) and it was deemed good because it had all the main gotchas detailed in it. Including the DevNonce issue when developing a device ...

**Reply by quintesse** [reply]
In respect to that old fa®t I had already read that doc from beginning to end but that doc itself doesn't talk about DevNonces, right? And in the troubleshooting section of the Wiki it points to the only thing it says about result code -1116 is "Not all uplinks to LoRaWAN server result in a downlink. This is not a problem, but rather a common behavior in LoRaWAN", which sounds like "don't worry about it" but then all the code examples treat it like a failure, so that confuses me. But I'll stop pestering you with questions... at least until I've RTFM.

**Reply by HeadBoffin** [reply]
https://github.com/jgromes/RadioLib/blob/577a0761ac0b5468c68ded009b77dc3460ed95dc/examples/LoRaWAN/LoRaWAN_Starter/notes.md?plain=1#L61

**Reply by HeadBoffin** [reply]
Once we've got you going, we'll have to take this and figure out the messages - every JR should result in a downlink.

For clarity, DevNonce is the technical name for "the join nonce"

**Reply by quintesse** [reply]
Just a small update. After having installed a new antenna on my device everything seems to working much better. The RSSI went up from -130 to -67 (without having moved it or the gateway) so that's good too.

I do get _some_ failed sends at times, but they actually arrive at the gateway, so perhaps the "ack" gets lost? 

Anyway, I need to start testing this out in the field to see if the signal will actually get through.
